keps/sig-storage/2317-fsgroup-on-mount/README.md

```diff
@@ -0,0 +1,400 @@
+-->
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: MountWithFSGroup
+    - Components depending on the feature gate:
+      - Kubelet
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Enabling this feature on its own does not changes any existing behaviour but if
+  selected CSI driver advertise `VOLUME_MOUNT_GROUP` node capability, it will result in
+  volume being mounted with pod's `fsGroup`.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes - feature gate can be disabled once enabled and kubelet will fallback to its current behavior, which will
+  result in any CSI driver that requires using `fsGroup` as a mount option to not use the mount option. In which case
+  mounted volumes will be unwritable by any pods other than those running as root.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  It will cause kubelet to use `volume_mount_group` field of CSI whenver applicable as discussed in above design.
+
+* **Are there any tests for feature enablement/disablement?**
+  Not yet.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+
+* **What specific metrics should inform a rollback?**
```

Comment:
I think you'll want to list the metrics that you use for determining successful and failed mounts here.  And probably count cases where the pod's intent wasn't fulfilled.

Reply:
Do we need these sections for alpha feature? The PRR check says:

> _This section must be completed when targeting beta graduation to a release._

Reply:
Updated this section. However kubelet never tracked whether pod's intent was fulfilled and I think such a metric is out of scope for this specific change because such a metric will apply to generic mechanism by which `fsGroup` is applied to the volumes (and they are in-use since k8s <1.2 or something)

Reply:
> Updated this section. However kubelet never tracked whether pod's intent was fulfilled and I think such a metric is out of scope for this specific change because such a metric will apply to generic mechanism by which `fsGroup` is applied to the volumes (and they are in-use since k8s <1.2 or something)

I can see an argument for not requiring it for alpha, but I don't think I follow the reasoning that because there was no metric before there shouldn't be a metric as this feature moves to beta.  Without a metric, how can you know that
1. the feature is in use
2. the feature is working.

Reply:
sure.  Added more information about metrics and filed https://github.com/kubernetes/kubernetes/issues/98667 to capture how `fsGroup` intent was full-filled.